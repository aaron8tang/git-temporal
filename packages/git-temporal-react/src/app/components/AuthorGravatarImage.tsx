import React from 'react';
import md5 from 'md5';
import { AuthorPlaceholderImage } from 'app/components/AuthorPlaceholderImage';
import { style } from 'app/styles';

export interface AuthorGravatarImageProps {
  width?: number;
  height?: number;
  emails?: string[];
  style?: object | string;
}

interface AuthorGravatarImageState {
  emailsIndex: number;
}

export class AuthorGravatarImage extends React.Component<
  AuthorGravatarImageProps,
  AuthorGravatarImageState
> {
  readonly state: AuthorGravatarImageState = {
    emailsIndex: 0,
  };

  componentDidUpdate(newProps) {
    newProps.emails.forEach((email, index) => {
      if (email !== this.props.emails[index]) {
        this.setState({ emailsIndex: 0 });
      }
    });
  }

  constructor(props) {
    super(props);

    this.onImageNotFound = this.onImageNotFound.bind(this);
  }

  render() {
    const { width = 75, height = 75, emails } = this.props;
    const { emailsIndex } = this.state;
    const notFound = emailsIndex >= emails.length;
    const outerStyle = style(this.props.style, {
      width,
      height,
      display: 'inline-block',
    });
    const gravatarEmail = (emails[emailsIndex] || '')
      .toLocaleLowerCase()
      .replace(/\s/g, '');
    const gravatarUrl = `https://www.gravatar.com/avatar/${md5(
      gravatarEmail
    )}?s=${width}&d=404`;

    const imageStyle = style({ width, maxHeight: height });
    return (
      <div style={outerStyle}>
        {notFound ? (
          <AuthorPlaceholderImage
            holderForEmails={emails}
            width={width}
            height={height}
          />
        ) : (
          <img
            data-for={emails[emailsIndex]}
            style={imageStyle}
            src={gravatarUrl}
            onError={this.onImageNotFound}
          />
        )}
      </div>
    );
  }

  onImageNotFound() {
    this.setState({ emailsIndex: this.state.emailsIndex + 1 });
  }
}
