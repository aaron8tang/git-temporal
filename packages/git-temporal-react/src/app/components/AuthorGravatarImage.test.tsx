import React from 'react';
import { shallow } from 'enzyme';

import { AuthorGravatarImage } from './AuthorGravatarImage';

describe('components/AuthorGravatarImage', () => {
  describe('when rendered with times in order', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(
        <AuthorGravatarImage emails={['bee.wilkerson@ymail.com']} />
      );
    });

    test('it should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
