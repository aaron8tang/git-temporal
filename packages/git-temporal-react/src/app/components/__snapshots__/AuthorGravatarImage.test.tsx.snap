// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AuthorGravatarImage when rendered with times in order it should match snapshot 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
      "height": 75,
      "width": 75,
    }
  }
>
  <img
    data-for="bee.wilkerson@ymail.com"
    onError={[Function]}
    src="https://www.gravatar.com/avatar/dabecaadcb5665b6ffec0712cdd3b0c5?s=75&d=404"
    style={
      Object {
        "maxHeight": 75,
        "width": 75,
      }
    }
  />
</div>
`;
